### DISCUSSION TITLE
Container Image not trusting LDAP certifcates

### DISCUSSION THREAD

**Discussion by ninjasftw** [apache/airflow#32109] [General]
### Apache Airflow version

2.6.2

### What happened

Setting up an Airflow instance which uses jumpcloud as its LDAP backend.

The underlying container(apache/airflow:latest-python3.10) does not trust the CA certficate at ldap.jumpcloud.com:636 even though it is a valid certificate

The webserver logs show 
`[2023-06-19T20:51:49.899+0000] {manager.py:1236} ERROR - {'result': -1, 'desc': "Can't contact LDAP server", 'ctrls': [], 'info': '(unknown error code)'}
`
digging into the underlying ldap / openssl functions seem to be where the issue is as you can trigger the error with 

```
> ldapsearch -d 9 -H 'ldaps://ldap.jumpcloud.com:636'
ldap_url_parse_ext(ldaps://ldap.jumpcloud.com:636)
ldap_create
ldap_url_parse_ext(ldaps://ldap.jumpcloud.com:636/??base)
ldap_pvt_sasl_getmech
ldap_search
put_filter: "(objectclass=*)"
put_filter: simple
put_simple_filter: "objectclass=*"
ldap_send_initial_request
ldap_new_connection 1 1 0
ldap_int_open_connection
ldap_connect_to_host: TCP ldap.jumpcloud.com:636
ldap_new_socket: 3
ldap_prepare_socket: 3
ldap_connect_to_host: Trying 3.70.144.176:636
ldap_pvt_connect: fd: 3 tm: -1 async: 0
attempting to connect: 
connect success
**TLS: peer cert untrusted or revoked (0x42)
TLS: can't connect: (unknown error code).**
ldap_msgfree
ldap_err2string
ldap_sasl_interactive_bind_s: Can't contact LDAP server (-1)
	additional info: (unknown error code)
```

The issue is not Airflow specific as such however its possible something in the Docker build process for the airflow image is doing something with the allowed ciphers or something

The container image is built from Debian 11 and a fresh Debian 11 instance does not show this issue.

as a workaround i've added a /etc/ldap/ldap.conf file into the container with the contents
`TLS_REQCERT never`

and that allows the ldapsearch command and webserver to talk to the LDAP instance

### What you think should happen instead

The ldap connection should be established.

Using a Debian11 container ldap search returns
```
ldapsearch -d 9 -H 'ldaps://ldap.jumpcloud.com:636'
ldap_url_parse_ext(ldaps://ldap.jumpcloud.com:636)
ldap_create
ldap_url_parse_ext(ldaps://ldap.jumpcloud.com:636/??base)
ldap_pvt_sasl_getmech
ldap_search
put_filter: "(objectclass=*)"
put_filter: simple
put_simple_filter: "objectclass=*"
ldap_send_initial_request
ldap_new_connection 1 1 0
ldap_int_open_connection
ldap_connect_to_host: TCP ldap.jumpcloud.com:636
ldap_new_socket: 3
ldap_prepare_socket: 3
ldap_connect_to_host: Trying 3.70.144.176:636
ldap_pvt_connect: fd: 3 tm: -1 async: 0
attempting to connect: 
connect success
```

### How to reproduce

`docker run --rm -it apache/airflow:2.6.0-python3.10 bash`

`ldapsearch -d 9 -H 'ldaps://ldap.jumpcloud.com:636'`

### Operating System

Debian GNU/Linux 11 (bullseye)

### Versions of Apache Airflow Providers

simply the base container image

### Deployment

Docker-Compose

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template! If you are willing to raise PR to address this issue please do so, no need to wait for approval.

**Comment by potiuk** [comment]
It's something you need to investigate and show. There is no guarantee that the "reference" image of Airflow will be 100% working if the server it connects to has some specific requirements. You are always free to extend the image https://airflow.apache.org/docs/docker-stack/build.html#extending-the-image as you need.

https://serverfault.com/questions/1115147/openldap-certificates-untrusted-or-revoked gives some hints and I suggest to do more search and finding how to solve your particular problem, but I am afraid it is not Airflow issue.

Remember: Reference images are a good start. And it's up to those who deploy it to solve some specific-to-their-deployment problems on top of it. Airflow is run as a group of volunteers (2500 of them) and if your business crtically depend on your LDAP backend specific requirements, maybe that's a good time to consider a paid support (Airflow is free, and the volunteers spent their free - weekend/evening time on helping on the premise they know and can help - this seems to be quite beyond of that).

**Reply by ninjasftw** [reply]
Hi @potiuk 
I am using the extended image which is where I saw the issue.  

I've also build a fresh image from a base Debian 11 container, installed Airflow and it works as expected.
My understanding was that the image provided was build from a Debian 11 container similar to what I built however for some reason the official image does not trust the certificate.  
My thought was that something in the airflow container build process for apache/airflow:latest etc made some changes that stopped it from trusting the presented certificate.

**Reply by potiuk** [reply]
Hard to say - we are using python base images (see https://github.com/apache/airflow/blob/main/Dockerfile#L49) "python:3.8-slim-bullseye" - official python images.

 if you can look at the differences there maybe  can figure out the differences. Maybe start from the same python images and add the stuff you want incrementally. Hard to say. There are many things that can be different between the "plain" debian images and "python ones" for sure we do not modify certificates in the image, maybe some extra packages need to be installed that are not present in the official python images.

**Reply by potiuk** [reply]
BTW. Likely you need to add some certicates (maybe Python devs did not like some), but what they are - hard to say. alternative is to change signing authority to be better supported for your certificate.
